### PR TITLE
[SPARK-34249][DOCS] Add documentation for ANSI implicit cast rules

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Add documentation for the ANSI implicit cast rules which are introduced from https://github.com/apache/spark/pull/31349

### Why are the changes needed?

Better documentation.

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

Build and preview in local:
image
image

image
image
image

### CODE REVIEW

**Comment by gengliangwang** [comment]
@srielau Thanks for summary the rules. He should be the first author on merging this PR.

File: docs/sql-ref-ansi-compliance.md [hunk 1]
```diff
@@ -160,6 +160,81 @@ SELECT * FROM t;
...
+| Data type | precedence list(from narrowest to widest)                        |
+|-----------|------------------------------------------------------------------|
+| Byte      | Byte -> Short -> Int -> Long -> Decimal -> Float* -> Double      |
+| Short     | Short -> Int -> Long -> Decimal-> Float* -> Double               |
+| Int       | Int -> Long -> Decimal -> Float* -> Double                       |
+| Long      | Long -> Decimal -> Float* -> Double                              |
+| Decimal   | Decimal -> Float* -> Double                                      |
+| Float     | Float -> Double                                                  |
+| Double    | Double                                                           |
+| Date      | Date->  Timestamp                                                |
```
Comment (MaxGekk) [inline docs/sql-ref-ansi-compliance.md:177]: `Date->  Timestamp` -> `Date -> Timestamp`

File: docs/sql-ref-ansi-compliance.md [hunk 2]
```diff
@@ -160,6 +160,81 @@ SELECT * FROM t;
...
+| Timestamp | Timestamp                                                        |
+| String    | String                                                           |
+| Binary    | Binary                                                           |
+| Boolean   | Boolean                                                          |
+| Interval  | Interval                                                         |
```
Comment (MaxGekk) [inline docs/sql-ref-ansi-compliance.md:182]: Should we distinguish year-month and day-time interval types?

File: docs/sql-ref-ansi-compliance.md [hunk 3]
```diff
@@ -160,6 +160,81 @@ SELECT * FROM t;
...
+
+This is a graphical depiction of the precedence list as a directed tree:
+<img src="img/type-precedence-list.png" width="80%" title="Type Precedence List" alt="Type Precedence List">
+
+#### Least Common Type Resolution
+The least common type from a set of types is the narrowest type reachable from the precedence list by all elements of the set of types.
+
+The least common type resolution is used to:
+- Decide whether a function expecting a parameter of a type can be invoked using an argument of a narrower type.
+- Derive the argument  type for functions which expect a shared argument type for multiple parameters, such as coalesce, least, or greatest.
```
Comment (MaxGekk) [inline docs/sql-ref-ansi-compliance.md:202]: Remove a space in `argument  type` -> `argument type`

File: docs/sql-ref-ansi-compliance.md (hunk 1 above)
Comment (gengliangwang) [reply docs/sql-ref-ansi-compliance.md]: Done

File: docs/sql-ref-ansi-compliance.md (hunk 3 above)
Comment (gengliangwang) [reply docs/sql-ref-ansi-compliance.md]: Done(BTW the extra space will be omitted in web page anyway)

File: docs/sql-ref-ansi-compliance.md (hunk 2 above)
Comment (gengliangwang) [reply docs/sql-ref-ansi-compliance.md:182]: Let's keep it simple in this section

**Review by MaxGekk — APPROVED** [review]

File: docs/sql-ref-ansi-compliance.md [hunk 4]
```diff
@@ -66,11 +66,9 @@ SELECT abs(-2147483648);
...
 ```
 
-### Type Conversion
+### Cast
 
-Spark SQL has three kinds of type conversions: explicit casting, type coercion, and store assignment casting.
 When `spark.sql.ansi.enabled` is set to `true`, explicit casting by `CAST` syntax throws a runtime exception for illegal cast patterns defined in the standard, e.g. casts from a string to an integer.
-On the other hand, `INSERT INTO` syntax throws an analysis exception when the ANSI mode enabled via `spark.sql.storeAssignmentPolicy=ANSI`.
 
 The type conversion of Spark ANSI mode follows the syntax rules of section 6.13 "cast specification" in [ISO/IEC 9075-2:2011 Information technology — Database languages - SQL — Part 2: Foundation (SQL/Foundation)](https://www.iso.org/standard/53682.html), except it specially allows the following
```
Comment (cloud-fan) [inline docs/sql-ref-ansi-compliance.md:73]: `The type conversion` -> `Explicit cast`

File: docs/sql-ref-ansi-compliance.md [hunk 5]
```diff
@@ -66,11 +66,9 @@ SELECT abs(-2147483648);
 +----------------+
 ```
 
-### Type Conversion
+### Cast
```
Comment (cloud-fan) [inline docs/sql-ref-ansi-compliance.md:69]: In this section, it says `In future releases, the behaviour of type coercion might change along with the other two type conversion rules.`

We should update it.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2!

**Pull request closed without merging**